Following support is added

srikrishnagopu opened the pull request:
1. cpu port to port map
2. adding udp port to vxlan
3. vxlan and qinq packet wrapper
4. few more verify_packet functionalities

antoninbas commented on platforms/eth.py:
```diff
@@ -25,4 +25,7 @@ def platform_config_update(config):
             (0, 3): 'veth7',
         }
 
+    cpu_port = config["cpu_port"]
```
This is too specific and should not be part of the core code. Please use this feature to specify the CPU port veth: https://github.com/p4lang/ptf#passing-parameters-to-tests

antoninbas commented on ptf:
```diff
@@ -66,6 +66,7 @@ config_default = {
     "platform_args"      : None,
     "platform_dir"       : os.path.join(root_dir, "platforms"),
     "interfaces"         : [],
+    "device_type"        : "bm",
```
Same remark as above, should be done with this: https://github.com/p4lang/ptf#passing-parameters-to-tests

antoninbas commented on src/ptf/dataplane.py:
```diff
@@ -362,7 +362,6 @@ def packets(self, device, port=None):
             if rcv_port == None:
                 self.logger.debug("Out of packets on all ports")
                 break
-
```
Why changes in this file?

antoninbas commented on src/ptf/testutils.py:
```diff
@@ -479,6 +480,83 @@ def simple_vxlan_packet(pktlen=300,
+              scapy.IPv6(src=ipv6_src, dst=ipv6_dst, fl=ipv6_fl, tc=ipv6_tc, hlim=ipv6_hlim)/ \
+            udp_hdr
+
+    if udp_dport == 4789:
```
VLAN_GPE is not part of ptf, please remove this block

antoninbas commented on src/ptf/testutils.py:
```diff
@@ -1406,4 +1540,104 @@ def verify_packets_any(test, pkt, ports=[], device_number=0):
+    verify_no_other_packets(test)
+
+    test.assertTrue(received == True, "Did not receive pkt on any of ports %r for device %d" % (ports, device_number))
+    return match_index
```
explain what match_index is in function docstring

antoninbas commented on src/ptf/testutils.py:
```diff
@@ -1406,4 +1540,104 @@ def verify_packets_any(test, pkt, ports=[], device_number=0):
+    for port in ports:
+        if rcv_device != device_number:
+            continue
+        if port in ports:
```
I don't understand this test. Your loop is "for port in ports", so isn't this test always going to be True?

antoninbas commented on src/ptf/testutils.py:
```diff
@@ -1406,4 +1540,104 @@ def verify_packets_any(test, pkt, ports=[], device_number=0):
+        timeout=1,
+        filters=FILTERS
+    )
+    for port in ports:
```
Can't you rewrite this in a more elegant way. Correct me if I'm wrong but I think all you are checking is "rcv_port in ports" and "rcv_pkt in pkts".

antoninbas commented on src/ptf/testutils.py:
```diff
@@ -1406,4 +1540,104 @@ def verify_packets_any(test, pkt, ports=[], device_number=0):
+    pkt_cnt = 0
+    test.assertTrue(len(pkts) == len(ports), "packet list count does not match port list count")
+    for port in ports:
+        if port in ports:
```
hum, same comment as above for this test?

antoninbas commented on src/ptf/testutils.py:
```diff
@@ -1406,4 +1540,104 @@ def verify_packets_any(test, pkt, ports=[], device_number=0):
+    received = False
+    index = 0
+    match_index = 0
+    (rcv_device, rcv_port, rcv_pkt, pkt_time) = test.dataplane.poll(
```
I think you want to use the dp_poll wrapper function here, no?

antoninbas commented on src/ptf/testutils.py:
```diff
@@ -1406,4 +1540,104 @@ def verify_packets_any(test, pkt, ports=[], device_number=0):
+    test.assertTrue(received == True, "Did not receive pkt on any of ports %r for device %d" % (ports, device_number))
+    return match_index
+
+def verify_packets_on_ports(test, pkts=[], ports=[], device_number=0):
```
this is not a very good name for this function, what about verify_each_packet_on_each_port?

antoninbas commented on src/ptf/testutils.py:
```diff
@@ -1406,4 +1540,104 @@ def verify_packets_any(test, pkt, ports=[], device_number=0):
+        filters=FILTERS
+    )
+    for port in ports:
+        if rcv_device != device_number:
```
instead of having this test, the device_number should be passed as an argument to dp_poll

antoninbas commented on src/ptf/testutils.py:
```diff
@@ -1406,4 +1540,104 @@ def verify_packets_any(test, pkt, ports=[], device_number=0):
+                port_number=port,
+                exp_pkt=pkt
+            )
+            if rcv_device != device_number:
```
this should be useless, the device_number was passed as an argument to dp_poll

antoninbas commented on src/ptf/testutils.py:
```diff
@@ -1406,4 +1540,104 @@ def verify_packets_any(test, pkt, ports=[], device_number=0):
+    verify_no_other_packets(test)
+    test.assertTrue(pkt_cnt == len(ports), "Did not receive pkt on one of ports %r for device %d" % (ports, device_number))
+
+def verify_any_packet_on_ports_list(test, pkts=[], ports=[], device_number=0):
```
I think this function is too much. I can live with the 2 above functions, but we're getting too specific here. This does not belong in the PTF core code.

antoninbas commented on src/ptf/testutils.py:
```diff
@@ -1406,4 +1540,104 @@ def verify_packets_any(test, pkt, ports=[], device_number=0):
+    verify_no_other_packets(test)
+    test.assertTrue(pkt_cnt == len(ports), "Did not receive pkt on one of ports %r for device %d" % (ports, device_number))
+
+def verify_packet_len(test, pkt, port, len, device_number=0):
```
Please rename to verify_packet_prefix

antoninbas commented on src/ptf/thriftutils.py:
```diff
@@ -16,6 +16,10 @@ def hex_to_byte(h):
 def uint_to_i32(u):
     if (u > 0x7FFFFFFF): u-= 0x100000000
     return u
+def byte_to_u(x):
```
char_to_uchar

antoninbas commented on src/ptf/thriftutils.py:
```diff
@@ -16,6 +16,10 @@ def hex_to_byte(h):
+def byte_to_u(x):
+    if (x >= 0):
+        return x
+    return 256+x
```
256 + x

srikrishnagopu commented:
fixed all the code review comments. 
1. removed vxlan gpe from testutils
2. removed cpu port and passing it as part of interface
3. using testparams for device type
4. did all renaming.

antoninbas commented on src/ptf/testutils.py:
```diff
@@ -1406,4 +1536,72 @@ def verify_packets_any(test, pkt, ports=[], device_number=0):
+            if str(pkt) == str(rcv_pkt):
+                match_index = ports.index(rcv_port)
+                received = True
+    verify_no_other_packets(test)
```
the call to verify_no_other_packets should include the device number

antoninbas commented on src/ptf/testutils.py:
```diff
@@ -1406,4 +1536,72 @@ def verify_packets_any(test, pkt, ports=[], device_number=0):
+    """
+    pkt_cnt = 0
+    test.assertTrue(len(pkts) == len(ports), "packet list count does not match port list count")
+    for device, port in ptf_ports():
```
is there a reason why you don't simply loop over the ports argument? This function is for a given device only anyway.
would this code work:
```
for port, pkt in zip(ports, pkts):
    // call to dp_poll
    test.assertTrue(rcv_pkt != None, ...)

verify no_other_packets(test, device_number)
```
I think the zip() call fits perfectly here.

antoninbas commented on src/ptf/testutils.py:
```diff
@@ -1406,4 +1536,72 @@ def verify_packets_any(test, pkt, ports=[], device_number=0):
+            if rcv_pkt != None:
+                pkt_cnt += 1
+
+    verify_no_other_packets(test)
```
need to include the device_number

srikrishnagopu commented:
done.

antoninbas commented on src/ptf/testutils.py:
```diff
@@ -1406,4 +1536,68 @@ def verify_packets_any(test, pkt, ports=[], device_number=0):
+            port_number=port,
+            exp_pkt=pkt
+        )
+        if rcv_pkt != None:
```
As I said in my previous comment, I think this should be removed and the assertTrue should be moved here:
```
test.assertTrue(rcv_pkt != None, ....)
```
This way we can know exactly which packet was not received instead of having a count.

srikrishnagopu commented:
done.

antoninbas commented:
ok thanks, I'll merge now
